proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,946 @@
+most aspects of type checking.
+
+Linear types also affect Core: Core must handle linear types in order
+to ensure that core-to-core passes do not break the linearity
+guarantees. The flip side is that all core-to-core passes must make
+sure that they do not break linearity. It is possible that some of the
+pre-linear-type passes actually do break linearity in some cases (this
+has not been acertained, yet).
+
+Unification of multiplicity expressions (as for for instance in the
+type of ``(.)`` above) requires some flavour of unification module
+associativity and commutativity (AC). Unification modulo AC is
+well-understood an relatively easy to implement. But would still be a
+non-trivial addition to the type-checker. We may decide that a
+simplified fragment is better suited for our use-case that the full
+generality of AC.
+
+
+Alternatives
+------------
+
+Syntax of multiplicity-parametric arrow
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The proposed mixfix ``a :p-> b`` syntax for the
+multiplicity-parametric arrow makes a potentially non-trivial addition
+to the parser. So does the proposed type constructor indexed by a
+multiplicity ``(:p->)``.
+
+A way to simplify the changes to the parser would be to have the type
+constructor be
+
+::
+
+  ``ARROW :: Multiplicity -> * -> * -- ignoring levitiy
+
+It would be very inconvenient to use a prefix notation for
+multiplicity parametric arrows: we wouldn't want the type of ``map``
+to read
+
+::
+
+  map :: ARROW 'U (ARROW p a b) (ARROW p [a] [b])
+
+So we introduce a binary type construction ``WithMult`` (or some
+operator syntax). It is a syntax error to use ``WithMult`` anywhere
+except to the left of an arrow. And ``WithMult a p -> b`` means
+``ARROW p a b``. So that the type of ``map`` becomes:
+
+::
+
+  map :: (a `WithMult` p -> b) -> [a] `WithMult` p -> [b]
+
+Subtyping instead of polymorphism
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Since ``A ->. B`` is a strengthening of ``A -> B``, it is tempting to
+make ``A ->. B`` a subtype of ``A -> B``. But subtyping and polymorphism
+don't mesh very well, and would yield a significantly more complex
+solution.
+
+In general, subtyping and polymorphism are not comparable, and some
+examples will work better with one or the other. Therefore it makes
+sense to go for the simplest one.
+
+In this proposal
+
+::
+
+  f :: A ->. B
+
+  g :: A -> B
+  g = f
+
+is, in theory, ill-typed. But it would be a problem to reject this
+program (especially with all the constructors which have been
+converted to linear types). So the type inference mechanism elaborates
+this program to the well-typed η-expansion
+
+::
+
+  f :: A ->. B
+
+  g :: A -> B
+  g x = f x
+
+No restriction on multiplicity variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Instead of restricting variables and type applications so that
+``case_p`` is allowed for a variable ``p``, we can allow arbitrary
+variables and disallow, in particular, ``case_p``.
```

Comment:
I think you will have to explain what disallowing `case_p` means. Perhaps that `a :p-> b` is allowed in types only if `p` is a multiplicity expression or a variable with a `CaseCompatible` constraint. I'm not sure I understand what you intend yet.